### PR TITLE
Get experiment data files from Google Drive format

### PR DESCRIPTION
**Pull request by jason-curtis**
[asana](https://app.asana.com/0/819671808102776/1147515614239923/f)

We talked about leaving this in a notebook for this ticket but it seemed obvious enough where it should go in a repo.

## Manual testing

Used on all the Drive directories in the log and eyeballed the results. see my fork of the Parse and combine calibration and sensor data snippet: http://marvin:8889/notebooks/Tools/Jupyter%20Snippets/Parse%20and%20combine%20calibration%20and%20sensor%20data-Temp%20version%20updated%20with%20Jason%20code.ipynb

this is what it took to tack the files on to a parsed data collection log:
```
from osmo_jupyter.dataset.files import get_experiment_data_files_by_type

DRIVE_EXPERIMENTS_ROOT = "/data/Google Drive/Technical/Experiments"
drive_directory_paths = parsed_log["drive_directory"].apply(
    lambda directory_name: os.path.join(DRIVE_EXPERIMENTS_ROOT, directory_name)
)
parsed_log_with_file_paths = pd.concat(
    [
        parsed_log, # parsed using `parse_data_collection_log`
        drive_directory_paths.apply(get_experiment_data_files_by_type).add_suffix(
            "_files"
        ),
    ],
    axis="columns",
)
```
image

*suggestion: I should probably go ahead and librarify this as well*, probably as `osmo_jupyter.dataset.parse.parse_data_collection_log_with_drive_files(log_filepath, drive_experiments_root)` but I'm not sure how much to pull on this scope creep thread before I'm stepping on Evan's ticket's toes! The logical next step after that would be to modify/provide a version of `open_and_combine_and_filter_source_data()` that's integrated with this also.

### CODE REVIEW

**Review by EvanSimpson — APPROVED** [review]
I'd be fine with pushing the scope creep back and librarifying the usage of this code in a future ticket, once we're ready to build the end-to-end pipeline.

File: osmo_jupyter/dataset/files.py [hunk 1]
```diff
@@ -0,0 +1,54 @@
+import os
+import pandas as pd
+
+FILE_TYPES = [
```
Comment (jaimemarijke) [inline osmo_jupyter/dataset/files.py]: are each of these their own folder? that's not what I expected from the name `FILE_TYPES`, but that's what it seems like they must be based on line 18

File: osmo_jupyter/dataset/files_test.py [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+    data_dir = os.path.join(parent_dir, "data")
+    os.mkdir(data_dir)
+
+    for relative_directory_path in directories_to_include:
+        os.makedirs(os.path.join(data_dir, relative_directory_path))
+
+    for relative_file_path in files_to_include:
+        with open(os.path.join(data_dir, relative_file_path), "w"):
+            # Just opening the file is enough to create it.
+            pass
```
Comment (jaimemarijke) [inline osmo_jupyter/dataset/files_test.py]: I'm vaguely remembering some new function that does this more directly

Comment (jaimemarijke) [reply osmo_jupyter/dataset/files_test.py]: ah, Path.touch(), e.g. 

```
from pathlib import Path
Path('/path/to/file.txt').touch()
```

not sure if I like that better

File: osmo_jupyter/dataset/files_test.py [hunk 3]
```diff
@@ -0,0 +1,87 @@
+import os
+import pandas as pd
+
+import osmo_jupyter.dataset.files as module
+
+
+def _init_data_dir(parent_dir, directories_to_include, files_to_include):
+    """set up a data directory with folders and files in it
+    Args:
```
Comment (jaimemarijke) [inline osmo_jupyter/dataset/files_test.py]: nit: missing `parent_dir` here

File: osmo_jupyter/dataset/files_test.py [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+            tmp_path, "this does NOT exist as a subdirectory"
+        )
+        expected = []
+        assert actual == expected
+
+    def test_gets_file_from_subdirectory_and_ignores_sub_subdirectories(self, tmp_path):
+        _init_data_dir(
+            tmp_path,
+            directories_to_include=["ysi_proodo/subdirectory"],
+            files_to_include=["ysi_proodo/somethin"],
```
Comment (jaimemarijke) [inline osmo_jupyter/dataset/files_test.py]: nit: could add a .txt extension or something to make it clearer that this is a file.

File: osmo_jupyter/dataset/files.py (hunk 1 above)
Comment (jaimemarijke) [reply osmo_jupyter/dataset/files.py]: maybe this could just use a comment with an example folder structure

File: osmo_jupyter/dataset/files.py [hunk 5]
```diff
@@ -0,0 +1,54 @@
+import os
```
Comment (jaimemarijke) [inline osmo_jupyter/dataset/files.py]: nit: the name `files.py` seems a little vague. maybe `raw_files` or `get_files`?

**Review by jaimemarijke — APPROVED** [review]

**Review by rush340 — APPROVED** [review]

File: osmo_jupyter/dataset/files.py (hunk 1 above)
Comment (rush340) [reply osmo_jupyter/dataset/files.py]: I agree that the `FILE_TYPES` name is a little confusing

File: osmo_jupyter/dataset/files_test.py (hunk 2 above)
Comment (jason-curtis) [reply osmo_jupyter/dataset/files_test.py]: holy mackerel. this is the library I always wanted but never knew I did. I spent altogether too much time in the last few workdays typing out `os.path.join()` calls.
prepare for a changeset that is overly psyched about this library

File: osmo_jupyter/dataset/files.py (hunk 5 above)
Comment (jason-curtis) [reply osmo_jupyter/dataset/files.py]: renamed to source_files

File: osmo_jupyter/dataset/files.py (hunk 1 above)
Comment (jason-curtis) [reply osmo_jupyter/dataset/files.py]: renamed to `FILE_TYPE_SUBFOLDERS` and added some documentation

**Comment by jason-curtis** [comment]
I switched to PosixPaths for everything. upside: that interface is much nicer to use than `os.path` when doing anything file-y. And in most cases it is backwards compatible. I think it's good in this case because we're expecting to do stuff programmatically with these paths almost immediately downstream.

Downside: when you save the output DataFrame as a CSV, the paths are wrapped in `PosixPath(...)` which would complicate deserialization. But you'd already have issues serializing/deserializing the lists that are in here. I think that's OK but we'll have to be aware of it if we ever try to serialize/deserialize this.

File: osmo_jupyter/dataset/source_files_test.py [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+        assert actual == expected
+
+    def test_gets_file_from_subdirectory_and_ignores_sub_subdirectories(self, tmp_path):
+        _init_data_dir(
+            tmp_path,
+            directories_to_include=["ysi_proodo/subdirectory"],
+            files_to_include=["ysi_proodo/somethin.csv"],
+        )
+        actual = module._get_experiment_data_file_paths_for_type(tmp_path, "ysi_proodo")
+        expected = [tmp_path / "data/ysi_proodo/somethin.csv"]
```
Comment (jaimemarijke) [inline osmo_jupyter/dataset/source_files_test.py]: nit: weird mix of Path and string

File: osmo_jupyter/dataset/source_files_test.py [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+            tmp_path, "this does NOT exist as a subdirectory"
+        )
+        expected = []
+        assert actual == expected
+
+    def test_gets_file_from_subdirectory_and_ignores_sub_subdirectories(self, tmp_path):
+        _init_data_dir(
+            tmp_path,
+            directories_to_include=["ysi_proodo/subdirectory"],
+            files_to_include=["ysi_proodo/somethin.csv"],
```
Comment (jaimemarijke) [inline osmo_jupyter/dataset/source_files_test.py]: also these aren't Paths

**Review by jaimemarijke — APPROVED** [review]
I'm okay with the change to using pathlib. not totally convinced by it yet but definitely seems worth trying. my only request would be to go 100% in that direction rather than mixing in string paths

**Review by rush340 — APPROVED** [review]
pathlib seems nice!
🚢

**Review by EvanSimpson — APPROVED** [review]

**Pull request merged**